File: 동규/1463.js
// X가 3으로 나누어 떨어지면, 3으로 나눈다. -> X % 3 == 0 : X / 3 + 1
// X가 2로 나누어 떨어지면, 2로 나눈다. -> X % 2 == 0 : X / 2 + 1
// 1을 뺀다.

const fs = require("fs");
const input = fs.readFileSync("example.txt");

const num = Number(input);
const dp = Array(num + 1).fill(0);

for (let i = 2; i <= num; i++) {
  dp[i] = dp[i - 1] + 1;
  if (i % 2 === 0) {
    dp[i] = Math.min(dp[i], dp[i / 2] + 1);
  }
  if (i % 3 === 0) {
    dp[i] = Math.min(dp[i], dp[i / 3] + 1);
  }
}

console.log(dp[num]);

File: 민성/1463.py
# BaekJoon 05/03 2023
# 1463 1로 만들기
# 13:35 ~ 13:51

N = int(input())

dp = [1e9] * (N+1) if N >= 3 else [1e9] * 4
dp[1] = 0
dp[2] = 1
dp[3] = 1

for i in range(4, N+1):
    if i % 3 == 0:
        dp[i] = min(dp[i], dp[i//3]+1)
    if i % 2 == 0:
        dp[i] = min(dp[i], dp[i//2]+1)
    dp[i] = min(dp[i], dp[i-1]+1)

print(dp[N])
